vm/sdk/wallet/tx.go

```diff
@@ -25,61 +28,69 @@ func encode(fields ...scale.Encodable) []byte {
 	return buf.Bytes()
 }
 
-// SelfSpawn creates a self-spawn transaction.
-func SelfSpawn(pk signing.PrivateKey, nonce core.Nonce, opts ...sdk.Opt) []byte {
-	// self-spawn has not yet been implemented for Athena
-	panic("self-spawn not yet implemented")
-}
-
 // Spawn creates a spawn transaction.
 func Spawn(
 	pk signing.PrivateKey,
-	template core.Address,
-	args scale.Encodable,
 	nonce core.Nonce,
 	opts ...sdk.Opt,
-) []byte {
+) ([]byte, error) {
 	options := sdk.Defaults()
 	for _, opt := range opts {
 		opt(options)
 	}
 
-	payload := core.Payload{}
-	payload.Nonce = nonce
-	payload.GasPrice = options.GasPrice
+	// Encode using the VM
+	vmlib, err := athcon.LoadLibrary(host.AthenaLibPath())
+	if err != nil {
+		panic(fmt.Errorf("loading Athena VM: %w", err))
+	}
+
+	meta := core.Metadata{}
+	meta.Nonce = nonce
+	meta.GasPrice = options.GasPrice
 
-	public := &core.PublicKey{}
-	copy(public[:], signing.Public(pk))
 	// note that principal is computed from pk
-	principal := core.ComputePrincipal(wallet.TemplateAddress, public)
+	athenaPayload := vmlib.EncodeTxSpawn(athcon.Bytes32(signing.Public(pk)))
+	principal, err := core.ComputePrincipalFromPubkey(wallet.TemplateAddress, *signing.NewPublicKey(signing.Public(pk)))
+	if err != nil {
+		return nil, err
+	}
+	payload := core.Payload(athenaPayload)
 
-	// TODO(lane): fix encoding
-	tx := encode(&sdk.TxVersion, &principal, &template, &payload, args)
-	sig := ed25519.Sign(ed25519.PrivateKey(pk), core.SigningBody(options.GenesisID[:], tx))
-	return append(tx, sig...)
+	tx := encode(&sdk.TxVersion, &principal, &meta, &payload)
```

Comment:
`payload` is already scale-encoded `Payload` type. Passing it to` encode` encodes it for the 2nd time. I think this should be:
```suggestion
	tx := encode(&sdk.TxVersion, &principal, &meta)
	tx = append(tx, payload...)
```

Reply:
See https://github.com/athenavm/athena/issues/172#issuecomment-2450539035